ir/src/constraint_builder/integrity_constraints/mod.rs

```diff
@@ -23,33 +23,55 @@ impl ConstraintBuilder {
     ) -> Result<(), SemanticError> {
         match stmt {
             IntegrityStmt::Constraint(constraint) => {
-                let (constraint_expr, _, selectors) = constraint.into_parts();
+                let (constraint_expr, cc_context, selectors) = constraint.into_parts();
                 match constraint_expr {
                     ConstraintExpr::Inline(inline_constraint) => {
-                        let (lhs, rhs) = inline_constraint.into_parts();
-                        // add the left hand side expression to the graph.
-                        let lhs = self.insert_expr(lhs)?;
-
-                        // add the right hand side expression to the graph.
-                        let rhs = self.insert_expr(rhs)?;
-
-                        // add the selectors expression to the graph
-                        let selectors = if let Some(selectors) = selectors {
-                            Some(self.insert_expr(selectors)?)
+                        let constraints = if let Some(cc_context) = cc_context {
+                            let constraint_exprs = self.unfold_cc(
+                                &ConstraintExpr::Inline(inline_constraint),
+                                &cc_context,
+                            )?;
+                            constraint_exprs
+                                .into_iter()
+                                .map(|expr| {
+                                    if let ConstraintExpr::Inline(constraint) = expr {
+                                        constraint
+                                    } else {
+                                        unreachable!("Unexpected Evaluator variant")
+                                    }
+                                })
+                                .collect::<Vec<_>>()
```

Comment:
I think we can simplify this whole implementation quite a bit by moving the processing of a single constraint into a dedicated method, and then calling that inside the comprehension method for when there are comprehensions and directly here for when there is no comprehension.

Here's what I propose for the `Constraint` variant in this method:

```
            IntegrityStmt::Constraint(constraint) => {
                let (constraint_expr, cc_context, selectors) = constraint.into_parts();

                // add the selectors expression to the graph
                let selectors = if let Some(selectors) = selectors {
                    Some(self.insert_expr(selectors)?)
                } else {
                    None
                };

                if let Some(cc_context) = cc_context {
                    self.unfold_cc(&constraint_expr, &cc_context, selectors)?;
                } else {
                    match constraint_expr {
                        ConstraintExpr::Inline(inline_constraint) => {
                            self.process_integrity_constraint(inline_constraint, selectors)?;
                        }
                        ConstraintExpr::Evaluator(ev_call) => {
                            self.process_evaluator_call(ev_call)?;
                        }
                    }
                };
            }
```

Here's the helper method:
```
    fn process_integrity_constraint(
        &mut self,
        constraint: InlineConstraintExpr,
        selectors: Option<NodeIndex>,
    ) -> Result<(), SemanticError> {
        // get the left and right hand side of the constraint
        let (lhs, rhs) = constraint.into_parts();

        // add the left hand side expression to the graph.
        let lhs = self.insert_expr(lhs)?;

        // add the right hand side expression to the graph.
        let rhs = self.insert_expr(rhs)?;

        // merge the two sides of the expression into a constraint.
        let root = self.merge_equal_exprs(lhs, rhs, selectors);

        // get the trace segment and domain of the constraint
        // the default domain for integrity constraints is `EveryRow`
        let (trace_segment, domain) = self.graph.node_details(&root, ConstraintDomain::EveryRow)?;

        // save the constraint information
        self.insert_constraint(root, trace_segment.into(), domain)?;

        Ok(())
    }
```

Reply:
Thanks, done.